### PR TITLE
CDAP-19050 remove GCE calls when they are not needed

### PR DESCRIPTION
**Pull request by albertshau**
Refactored DataprocClient to use a new NetworkClient interface
to perform all operations that used to require a call to the
GCE APIs. A ComputeNetworkClient implementation contains all the
existing logic, while a NoOpNetworkClient implementation performs
all operations without making any external calls. When SSH is
required to monitor the program run, the ComputeNetworkClient is
used. When runtime monitoring is used, the NoOpNetworkClient is
used.

### CODE REVIEW

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+  private Network networkInfo;
+
+  public ComputeNetworkClient(DataprocConf conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public void init(@Nullable String networkName) throws RetryableProvisionException, IOException,
+    GeneralSecurityException {
+    compute = getCompute(conf);
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:83]: this content was copied from the latter half of DataprocClient.fromConf()

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+    return networkInfo.getSelfLink();
+  }
+
+  /**
+   * Lookup up internal/external IP addresses for the node and creation time of the node.
+   */
+  @Override
+  public Node getNode(Node.Type type, String zone, String nodeName) throws IOException {
+    Instance instance;
+    try {
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:130]: this method was moved from DataprocClient.getNode()

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    // For internal IP only cluster, nodes only have ip.internal.
+    String ip = properties.get("ip.external");
+    if (ip == null) {
+      ip = properties.get("ip.internal");
+    }
+    return new Node(nodeName, type, ip, ts, properties);
+  }
+
+  @Override
+  public List<String> getAdditionalTags(boolean isInternalIPOnly) throws RetryableProvisionException, IOException {
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java:174]: this method was moved from DataprocClient.getFirewallTags()

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 4]
```diff
@@ -191,134 +120,12 @@ private static DataprocClient getDataprocClient(DataprocConf conf, boolean requi
...
-      network = findNetwork(networkHostProjectId, compute);
-    }
-    if (network == null) {
-      throw new IllegalArgumentException("Unable to automatically detect a network, please explicitly set a network.");
     }
 
-    String subnet = conf.getSubnet();
-    Network networkInfo = getNetworkInfo(networkHostProjectId, network, compute);
+    NetworkClient networkClient = requireNetwork ? new ComputeNetworkClient(conf) : new NoOpNetworkClient();
+    networkClient.init(network);
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: the logic removed from this method was put into ComputeNetworkClient.init()

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 5]
```diff
@@ -191,134 +120,12 @@ private static DataprocClient getDataprocClient(DataprocConf conf, boolean requi
...
-                                                         network, networkHostProjectId));
-      }
-    }
-
-    subnet = chooseSubnet(network, subnets, subnet, conf.getRegion());
-
-    return new DataprocClient(new DataprocConf(conf, network, subnet), client, compute, networkInfo);
-  }
-
-  private static PeeringState getPeeringState(String systemProjectId, String systemNetwork, Network networkInfo) {
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:230]: everything else removed from this class was moved to ComputeNetworkClient without any changes.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 6]
```diff
@@ -191,134 +120,12 @@ private static DataprocClient getDataprocClient(DataprocConf conf, boolean requi
       // If the CDAP instance is running on a GCE/GKE VM from a project that matches the provisioner project,
       // use the network of that VM.
       network = systemNetwork;
-    } else if (network == null) {
-      // Otherwise, pick a network from the configured project using the Compute API
-
-      network = findNetwork(networkHostProjectId, compute);
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java:197]: this is the main functional change. This logic still happens in ComputeNetworkClient.init(), but not in NoOpNetworkClient(). NoOpNetworkClient is used when runtime monitoring is being used (in a GCP environment) instead of SSH monitoring (in non-GCP environment). 

If the dataproc network is not given, and the network cannot be auto-detected, we now throw an exception saying the network must be provided instead of making a GCE call to pick a network.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * A NetworkClient that is backed by actual calls to the GCP Compute APIs.
+ */
+public class ComputeNetworkClient implements NetworkClient {
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java]: basically the entirety of this class is moved from DataprocClient

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+      throw new IllegalArgumentException(String.format(
+        "Could not find an ingress firewall rule for network '%s' in project '%s' for ports '%s'. " +
+          "Please create a rule to allow incoming traffic on those ports for your IP range.",
+        networkInfo.getName(), conf.getNetworkHostProjectID(), portList));
+    }
+    return tags;
+  }
+
+  @Override
+  public boolean isNetworkPeered(String systemProjectId, String systemNetwork) {
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java]: this is one of the few parts that is not cut and paste from DataprocClient. This is taken from the end of DataprocClient.isInternalIPOnly()

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+  private Compute compute;
+  private DataprocConf conf;
+  private Network networkInfo;
+
+  public ComputeNetworkClient(DataprocConf conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public void init(@Nullable String networkName) throws RetryableProvisionException, IOException,
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java]: this is taken from the latter half of DataprocClient.fromConf()

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocTool.java [hunk 10]
```diff
@@ -105,10 +106,11 @@ public static void main(String[] args) throws Exception {
       conf = DataprocConf.create(properties);
     }
 
-    String imageVersion = commandLine.hasOption('i') ? commandLine.getOptionValue('i') : "1.2";
+    String imageVersion = commandLine.hasOption('i') ? commandLine.getOptionValue('i') : "1.3";
```
Comment (fernst) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocTool.java]: Why don't we update this default to 2.0? 1.3 is currently unsupported by Dataproc.

**Review by fernst — COMMENTED** [review]
Asked a question. Other than that, code looks good.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+      throw new IllegalArgumentException(String.format(
+        "Could not find an ingress firewall rule for network '%s' in project '%s' for ports '%s'. " +
+          "Please create a rule to allow incoming traffic on those ports for your IP range.",
+        networkInfo.getName(), conf.getNetworkHostProjectID(), portList));
+    }
+    return tags;
+  }
+
+  @Override
+  public boolean isNetworkPeered(String systemProjectId, String systemNetwork) {
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java]: Please add javadocs for all public methods

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 12]
```diff
@@ -103,73 +78,23 @@
...
-    RetryableProvisionException {
-    try {
-      return getDataprocClient(conf, requireNetwork);
-    } catch (HttpResponseException e) {
-      if (e.getStatusCode() == HttpStatusCodes.STATUS_CODE_SERVICE_UNAVAILABLE) {
-        throw new RetryableProvisionException(e);
-      }
-      throw e;
-    }
+  DataprocClient(DataprocConf conf, ClusterControllerClient client, NetworkClient nodeClient) {
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: Given factory method below, should this be private?

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 13]
```diff
@@ -436,22 +212,21 @@ ClusterOperationMetadata createCluster(String name, String imageVersion, Map<Str
       if (conf.getSubnet() != null) {
         clusterConfig.setSubnetworkUri(conf.getSubnet());
       } else {
-        clusterConfig.setNetworkUri(network.getSelfLink());
+        clusterConfig.setNetworkUri(nodeClient.getNetwork());
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: Any specific reason to name it `nodeClient` here and `networkClient` in `fromConf`?

Comment (tivv) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: Also as there are cases where we don't fill `network`, could you add assert on any `getNetwork` call to ensure we don't try to use it when it's not available?

**Review by tivv — COMMENTED** [review]
So, my high level question is: We use network name to create a cluster. Does it mean this PR changes it to null if we are not using SSH?

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocTool.java [hunk 14]
```diff
@@ -57,6 +57,7 @@ public static void main(String[] args) throws Exception {
       .addOption(new Option("p", "project", true, "Google cloud project id."))
       .addOption(new Option("c", "configFile", true, "File all provisioner settings as a json object."))
       .addOption(new Option("i", "imageVersion", true, "The image version for the cluster. Defaults to 1.2."))
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocTool.java]: Change default here are you are changing it below. And I agree that 2.0 should be default

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java [hunk 15]
```diff
@@ -419,7 +418,7 @@ protected String getClusterName(ProvisionerContext context) throws Exception {
   @VisibleForTesting
   protected DataprocClient getClient(DataprocConf conf) throws IOException, GeneralSecurityException,
     RetryableProvisionException {
-    return DataprocClient.fromConf(conf);
+    return DataprocClient.fromConf(conf, conf.getPublicKey() != null);
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java]: It would be great not to do those calls for deprovisioning as it's not needed irrespective of SSH or not

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java (hunk 13 above)
Comment (albertshau) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: yep, this was an old name I forgot to change, will make it consistent.

This will always be non-null, will add a comment on the relevant part

**Comment by albertshau** [comment]
Sorry I forgot to push latest changes, the network (for creation) is supposed to always be non-null. We still need to look it up when it is not given, even without SSH.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 16]
```diff
@@ -180,145 +105,31 @@ private static DataprocClient getDataprocClient(DataprocConf conf, boolean requi
...
-    for (String currentSubnet : subnets) {
-      // if a subnet name is given then get the region of that subnet based on the zone
-      if (subnet != null && !currentSubnet.endsWith("subnetworks/" + subnet)) {
-        continue;
-      }
-      if (currentSubnet.contains(region + "/subnetworks")) {
-        return currentSubnet;
-      }
+      // Otherwise, make some Compute calls to pick a network
+      networkClient.init(null);
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: over here, we still look up the network if null, even if SSH is not required.

I think I'll refactor this a bit, since we don't need to do this lookup for non-create operations.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java (hunk 15 above)
Comment (albertshau) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java]: good point, I'll make this change.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java (hunk 11 above)
Comment (albertshau) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeNetworkClient.java]: have added docs to the interface

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java [hunk 17]
```diff
@@ -1127,7 +710,7 @@ private io.cdap.cdap.runtime.spi.provisioner.ClusterStatus convertStatus(Cluster
   }
 
   @Override
-  public void close() {
+  public void close() throws Exception {
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java]: Was there `ClusterControllerClient` change?

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/Clients.java [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+      network = networkClient.getNetwork();
+    }
+
+    if (shouldUseSSH && networkClient != null) {
+      return networkClient;
+    }
+
+    networkClient = shouldUseSSH ?
+      new ComputeClient(conf, Clients.getCompute(conf)) : new NoOpClient();
+    networkClient.init(network, subnet);
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/Clients.java]: nit: combine this ternary with previous if. It would be easier to follow the logic.

File: cdap-runtime-ext-dataproc/src/test/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisionerTest.java [hunk 19]
```diff
@@ -288,11 +291,17 @@ public void testClusterCreateNoReuse() throws Exception {
     context.setSparkCompat(SparkCompat.SPARK2_2_11);
     context.addProperty(DataprocConf.CLUSTER_REUSE_ENABLED, "false");
 
+    ComputeClient computeClientMock = Mockito.mock(ComputeClient.class);
+    PowerMockito.spy(Clients.class);
+    PowerMockito.doReturn(computeClientMock)
+      .when(Clients.class, "createNetworkClient", Mockito.any(), Mockito.eq(true));
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/test/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisionerTest.java]: nit: Instead of mocking a static method it's often a good idea to use a factory class instead of factory method

**Review by tivv — COMMENTED** [review]
As of now the logic is pretty hard to follow, primarily because ComputeClient is doing two things: network detection & ssh-specific configuration data retrieval. Also ssh-specific part is split between DataprocClient  and NetworkClient and it's non-obvious of what would be  available for DataprocClient from NetworkClient depending on if SSH is needed or not.
I'd suggest to separate out network information / detection and ssh initialization into two separate entities (either unrelated or with inheritance). May be even make lazy network detection on on-needed basis. This should simplify a lot of "if else" code and organize all "if ssh" things in one place, making it much easier to comprehend

**Comment by albertshau** [comment]
I'll see what I can do. One of the difficulties is the result of the initial call to get the network information is used in all the methods

**Comment by tivv** [comment]
@albertshau how about doing it lazy on request? I think it would simplify things a lot. You can have just 1 implementation that just don't do this calls if they are not requested

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/SSHDataprocClient.java [hunk 20]
```diff
@@ -0,0 +1,414 @@
...
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Wrapper around the dataproc client that adheres to our configuration settings. Creates Dataproc clusters that
+ * are accessible through SSH.
+ */
+class SSHDataprocClient extends DataprocClient {
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/SSHDataprocClient.java:56]: this class is just code moved from DataprocClient verbatim

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/RuntimeMonitorDataprocClient.java [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+import com.google.cloud.dataproc.v1.ClusterControllerClient;
+import com.google.cloud.dataproc.v1.GceClusterConfig;
+import io.cdap.cdap.runtime.spi.provisioner.Node;
+
+import java.util.Collections;
+
+/**
+ * Wrapper around the dataproc client that adheres to our configuration settings.
+ */
+class RuntimeMonitorDataprocClient extends DataprocClient {
```
Comment (albertshau) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/RuntimeMonitorDataprocClient.java:29]: behavior changes are local to this class, where it doesn't lookup node information and it doesn't look at networking setup to determine whether the cluster should be private ip only

File: cdap-runtime-ext-dataproc/src/test/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClientTest.java [hunk 22]
```diff
@@ -61,13 +63,10 @@ public void init() throws Exception {
     properties.put("accountKey", "{ \"type\": \"test\"}");
     properties.put(DataprocConf.PROJECT_ID_KEY, "dummy-project");
     properties.put("zone", "us-test1-c");
-    dataprocConf = DataprocConf.create(properties, null);
+    dataprocConf = DataprocConf.create(properties);
 
-    PowerMockito.spy(DataprocClient.class);
```
Comment (tivv) [inline cdap-runtime-ext-dataproc/src/test/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClientTest.java:66]: Thanks! static method mocking is always a code smell

**Review by tivv — APPROVED** [review]

File: cdap-runtime-ext-dataproc/src/test/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClientTest.java (hunk 22 above)
Comment (albertshau) [reply cdap-runtime-ext-dataproc/src/test/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClientTest.java:66]: 👍  I didn't even know you could mock static methods until seeing this test
